authedclient.go

```diff
@@ -0,0 +1,140 @@
+package gomatrixserverlib
+
+import (
+	"encoding/json"
+	"github.com/matrix-org/gomatrix"
+	"golang.org/x/crypto/ed25519"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+)
+
+// An AuthedClient is a matrix federation client that adds
+// "Authorization: X-Matrix" headers to requests that need ed25519 signatures
+type AuthedClient struct {
+	Client
+	serverName       ServerName
+	serverKeyID      KeyID
+	serverPrivateKey ed25519.PrivateKey
+}
+
+// NewAuthedClient makes a new AuthedClient
+func NewAuthedClient(
+	serverName ServerName, keyID KeyID, privateKey ed25519.PrivateKey,
+) *AuthedClient {
+	return &AuthedClient{
+		Client:           Client{client: http.Client{Transport: newFederationTripper()}},
+		serverName:       serverName,
+		serverKeyID:      keyID,
+		serverPrivateKey: privateKey,
+	}
+}
+
+func (ac *AuthedClient) doRequest(r FederationRequest, resBody interface{}) error {
+	if err := r.Sign(ac.serverName, ac.serverKeyID, ac.serverPrivateKey); err != nil {
+		return err
+	}
+
+	req, err := r.HTTPRequest()
+	if err != nil {
+		return err
+	}
+
+	res, err := ac.client.Do(req)
+	if res != nil {
+		defer res.Body.Close()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	contents, err := ioutil.ReadAll(res.Body)
+	if res.StatusCode/100 != 2 { // not 2xx
+		// Adapted from https://github.com/matrix-org/gomatrix/blob/master/client.go
+		var wrap error
+		var respErr gomatrix.RespError
+		if _ = json.Unmarshal(contents, &respErr); respErr.ErrCode != "" {
+			wrap = respErr
+		}
+
+		// If we failed to decode as RespError, don't just drop the HTTP body, include it in the
+		// HTTP error instead (e.g proxy errors which return HTML).
+		msg := "Failed to " + r.Method() + " JSON to " + r.RequestURI()
+		if wrap == nil {
+			msg = msg + ": " + string(contents)
+		}
+
+		return gomatrix.HTTPError{
+			Code:         res.StatusCode,
+			Message:      msg,
+			WrappedError: wrap,
+		}
+	}
+
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(contents, resBody)
+}
+
+// SendTransaction sends a transaction
```

Comment:
Now if this were the CS API I'd be linking to the spec... but in absence of that, I'm going to need you to go into a bit more detail here.
 - When does this fail?
 - Under what failures is it safe to retry (transient failures)
 - Presumably the transaction must be immutable by this point: we cannot add more PDUs between retries.
 - Do we update `OriginServerTS` between retries?